src/peergos/shared/user/fs/FileWrapper.java

```diff
@@ -399,58 +405,44 @@ public boolean isDirty() {
     }
 
-    public CompletableFuture<FileWrapper> uploadFileSection(String filename, AsyncReader fileData, long startIndex, long endIndex,
-                                                            NetworkAccess network, SafeRandom random,
-                                                            ProgressConsumer<Long> monitor, Fragmenter fragmenter) {
-        return uploadFileSection(filename, fileData, startIndex, endIndex, Optional.empty(), true, network, random, monitor, fragmenter);
+    @JsMethod
+    public CompletableFuture<FileWrapper> uploadFileJS(String filename, AsyncReader fileData,
+                                                       int lengthHi, int lengthLow,
+                                                       boolean overwriteExisting,
+                                                       NetworkAccess network, SafeRandom random,
+                                                       ProgressConsumer<Long> monitor, Fragmenter fragmenter,
+                                                       List<Location> locations) {
+        return uploadFileSection(filename, fileData, false, 0, lengthLow + ((lengthHi & 0xFFFFFFFFL) << 32),
+                Optional.empty(), overwriteExisting, network, random, monitor, fragmenter, locations);
     }
 
-    public CompletableFuture<FileWrapper> uploadFileSection(String filename,
-                                                            AsyncReader fileData,
-                                                            long startIndex,
-                                                            long endIndex,
-                                                            Optional<SymmetricKey> baseKey,
-                                                            boolean overwriteExisting,
-                                                            NetworkAccess network,
-                                                            SafeRandom random,
-                                                            ProgressConsumer<Long> monitor,
-                                                            Fragmenter fragmenter) {
-        return uploadFileSection(filename, fileData, false, startIndex, endIndex, baseKey,
-                overwriteExisting, network, random, monitor, fragmenter);
+    public CompletableFuture<FileWrapper> uploadOrOverwriteFile(String filename,
+                                                                AsyncReader fileData,
+                                                                long length,
+                                                                NetworkAccess network,
+                                                                SafeRandom random,
+                                                                ProgressConsumer<Long> monitor,
+                                                                Fragmenter fragmenter,
+                                                                List<Location> locations) {
```

Comment:
Is it necessary to pass in ``locations``? 

Where it  is being called in the PR  the argument  could be replaced with``this.generateChildLocationsFromSize(length, random)`` AFAICT.

Reply:
in the pr yes, but the pr doesn't actually do any transactions. The parameter needs to be there for transactions.

Reply:
Not sure what you mean by shared storage, but probably yes.

Reply:
The real problem is principle of least privilege, because that overloaded method must have higher privilege than it needs.

Reply:
One solution to that would be to give the transaction folder a different signing key. That would almost minimise the excess privilege being passed around.